[JENKINS-36707] Add `CI` environment variable

usmonster opened the pull request:
See [JENKINS-36707](https://issues.jenkins-ci.org/browse/JENKINS-36707).

Introduces a standard `CI` env var that is always set to "true" in order
to allow pipelines and tools to know they're executing in a continuous
integration context. This has emerged as a best practice used by most
other popular CI solutions, and a number of tools look for this variable
in order to make CI-appropriate optimizations (cf. the ticket).

Closes JENKINS-36707.

### Proposed changelog entries

* JENKINS-36707, Adds a "CI" environment variable, always set to "true" as in most other CI solutions. Note that the presence of this variable may change behavior of some of the tools used in your pipelines.

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [x] (If applicable) Jira issue is well described
- [x] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [x] Appropriate autotests or explanation to why this change has no tests
- [x] For dependency updates: links to external changelogs and, if possible, full diffs _(N/A)_

### Desired reviewers

@daniel-beck (since he's active on the ticket, but anyone can review)

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins-ci.org/issues/?filter=12146)).

usmonster commented on core/src/main/java/jenkins/model/CoreEnvironmentContributor.java:
```diff
@@ -41,6 +41,8 @@ public void buildEnvironmentFor(Run r, EnvVars env, TaskListener listener) throw
 
     @Override
     public void buildEnvironmentFor(Job j, EnvVars env, TaskListener listener) throws IOException, InterruptedException {
+        env.put("CI", "true");
```
Please note: I purposely didn't update the [buildEnv.groovy](https://github.com/jenkinsci/jenkins/blob/master/core/src/main/resources/jenkins/model/CoreEnvironmentContributor/buildEnv.groovy) or [buildEnv.properties](https://github.com/jenkinsci/jenkins/blob/master/core/src/main/resources/jenkins/model/CoreEnvironmentContributor/buildEnv.properties) files since it's a "constant variable" and technically has no real semantic value. Itfyou think it's worth adding it, I can update this PR to do so.

oleg-nenashev replied:
Would be better to add it, because it might override user-defined variables

oleg-nenashev reviewed (commented):
Would it also make sense to set CI_NAME? Some projects set it, e.g. CodeShip: https://docs.cloudbees.com/docs/cloudbees-codeship/latest/pro-builds-and-configuration/environment-variables . The tricky point is that this field should be customizable with Jenkins, because Jenkins, Jenkinsfule Runner, DotCI and other impleentations might be completely different even if they use the Jenkins core

usmonster commented:
Thanks for the lightning-fast feedback!

> Would it also make sense to set CI_NAME?

I'm not sure it's needed right away based on the use cases described in the ticket, so I'd leave that for another ticket & PR.

usmonster committed:
[JENKINS-36707] Add `CI` environment variable

Introduces a standard `CI` env var that is always set to "true" in order
to allow pipelines and tools to know they're executing in a continuous
integration context. This has emerged as a best practice used by most
other popular CI solutions, and a number of tools look for this variable
in order to make CI-appropriate optimizations (cf. the ticket).

Closes JENKINS-36707.

daniel-beck commented:
> Would it also make sense to set CI_NAME? Some projects set it, e.g. CodeShip:

So any time any CI related tool sets any env var, we should do the same, or what's the argument here?

---

This looks reasonable. Must not be part of characteristic env vars, which it doesn't look like it is, so should be fine. I'm still not convinced but OTOH it's also fairly cheap to do, and generic.

res0nance reviewed (approved):
IMO Reasonable change, going through the docs of other CI tools they all seem to define
`CI=true` and `<CI_TOOL_NAME>=true` we don't know if we want the latter but the former seems ubiquitous.

In terms of tooling I'm not sure what automated behaviours CLI tools take when in a CI environment but this might help some of them (although if it does this can alter some behaviour as well).

oleg-nenashev reviewed (approved):
Yes, this pull request can be merged. Not blocking it due to the CI_NAME discussion

daniel-beck commented:
Just manually tested this. In freestyle jobs, an agent-wide environment variable takes precedence over this one (and a build parameter is highest), while in Pipelines, this one wins over agent env vars.

Also, given Jenkins's case-insensitive handling of environment variables, any existing build parameters `ci` (or any other case) will now be uppercased.

oleg-nenashev commented:
So, what do we want to do with that? Environment variable handling is generally a mess in Jenkins.
We could just add a disclaimer to the upgrade guide

usmonster commented:
Thanks for testing!

Is the behavior you describe any different from any of the other environment variables being set in the same function (`JENKINS_URL`, for example)?

Just trying to determine if this is a real concern, or if you're just making explicit the cases that make these breaking changes.

Thanks!

daniel-beck commented:
> Is the behavior you describe any different from any of the other environment variables being set in the same function (`JENKINS_URL`, for example)?

No. But they a) already exist and b) have a fairly unique name.

> Just trying to determine if this is a real concern, or if you're just making explicit the cases that make these breaking changes.

Depends on whether someone has a variable `ci` defined or not 😁 

My comment is non-blocking.

oleg-nenashev commented:
I will start the merge countdown. 

We may merge it in 24 hours if there is no negative feedback. Please see the [merge process documentation](https://github.com/jenkinsci/jenkins/blob/master/docs/MAINTAINERS.adoc#merge-process) for more information about the merge process
